packages/core/src/components/time-picker/time-picker.tsx

```diff
@@ -131,43 +202,248 @@ export class TimePicker {
+    const numberArray = this.getNumberArrayForUnit(this.focusedUnit);
+    const maxValue = numberArray[numberArray.length - 1];
+    const minValue = numberArray[0];
+
+    if (value > maxValue) {
+      value = minValue;
+    } else if (value < minValue) {
+      value = maxValue;
+    }
+
+    this.focusedValue = value;
+
+    const cellElement = this.hostElement.shadowRoot?.querySelector(
+      `[data-element-container-id="${this.focusedUnit}-${value}"]`
+    ) as HTMLDivElement;
+
+    if (cellElement) {
+      cellElement.setAttribute('tabindex', '0');
+      cellElement.focus();
+    }
+  }
+
+  private moveFocusToAdjacentUnit(direction: number) {
+    const currentUnitIndex = this.timePickerDescriptors.findIndex(
+      (d) => d.unit === this.focusedUnit
+    );
+    const newUnitIndex = !this.ampmFocus
+      ? currentUnitIndex + direction
+      : currentUnitIndex;
+
+    // Check if new index is within bounds
+    if (newUnitIndex >= 0 && newUnitIndex < this.timePickerDescriptors.length) {
+      const newUnit = this.timePickerDescriptors[newUnitIndex].unit;
+
+      const selectedValue = Number(this.formattedTime[newUnit]);
+
+      this.focusedUnit = newUnit;
+      this.updateFocusedValue(selectedValue);
+      this.ampmFocus = undefined;
+    } else if (
+      direction > 0 &&
+      this.timeRef !== undefined &&
+      newUnitIndex === this.timePickerDescriptors.length
+    ) {
+      // If moving right and we have AM/PM, focus that
+      this.handleAmpmFocus(this.timeRef);
+    }
+  }
+
+  private handleAmpmFocus(element: 'AM' | 'PM') {
+    this.ampmFocus = element;
+
+    const ampmContainer = this.hostElement.shadowRoot?.querySelector(
+      `[data-am-pm-id="${element}"]`
+    ) as HTMLDivElement;
+
+    if (ampmContainer) {
+      ampmContainer.setAttribute('tabindex', '0');
+      ampmContainer.focus();
+    }
+  }
+
+  private setupVisibilityObserver() {
+    let dropdownElement: Element | null = this.hostElement;
+    while (dropdownElement && dropdownElement.tagName !== 'IX-DROPDOWN') {
+      dropdownElement = dropdownElement.parentElement;
+    }
+
+    if (!dropdownElement) {
+      return;
+    }
+
+    this.visibilityObserver = new MutationObserver((mutations) => {
+      for (const mutation of mutations) {
+        if (mutation.type === 'attributes') {
+          const dropdown = mutation.target as HTMLElement;
+
+          // Check if the dropdown is now visible (has the 'show' class)
+          if (dropdown.classList.contains('show')) {
+            // Check if elements are fully rendered by looking at their dimensions
+            const elementsReady = this.areElementsRendered();
+
+            if (elementsReady && !this.hasUpdatedScrollPositions) {
+              this.updateScrollPositions();
+              this.hasUpdatedScrollPositions = true;
+            }
+          } else {
+            this.hasUpdatedScrollPositions = false;
+          }
+        }
+      }
+    });
```

Comment:
Extract to a function would make it more readable. Also we could here also a `Guard Clauses` 

```
if (mutation.type !== 'attribute') {
   continue;
}
```

to improve the readability.